### PR TITLE
[WIP] Resolve overloads through the constraint system

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch tracks the rewrite of the overload resolution engine in terms of our new constraint solver.

Ideally, this overload resolver will punish coercions in order to arrive at the most optimal overload for a given function call.

### CODE REVIEW

**Comment by harlanhaskins** [comment]
Aw yeah, the build is super broken right now! 😄

**Comment by segiddins** [comment]
Lol I'm guessing you broke compiling the stdlib

**Comment by harlanhaskins** [comment]
All overload resolution is broken because the constraint generator is relying on decls being wired up. Not sure how to fix this just yet...

File: Sources/Base/ASTContext.swift [hunk 1]
```diff
@@ -224,83 +224,6 @@ public class ASTContext {
     return missing
   }
 
-  /// - Returns: Whether the expression's type was changed
-  @discardableResult
```
Comment (CodaFi) [inline Sources/Base/ASTContext.swift]: Praise be, praise be

File: Sources/Base/Types.swift [hunk 2]
```diff
@@ -10,38 +10,101 @@ enum FloatingPointType {
...
+  /// A data type representing a string literal that hasn't been resolved to
+  /// a concrete String or *Int8 type.
+  case stringLiteral
+
+  /// A data type representing an integer literal that hasn't been resolved to
+  /// a concrete integer type.
+  case integerLiteral
+
+  /// A data type representing a `nil` literal that hasn't been resolved to
+  /// a concrete pointer type.
```
Comment (segiddins) [inline Sources/Base/Types.swift]: `indirect` or `reference`

File: Sources/Base/Types.swift [hunk 3]
```diff
@@ -10,38 +10,101 @@ enum FloatingPointType {
...
+  case nilLiteral
+
+  /// A function from <params> -> <returnType>
   indirect case function(args: [DataType], returnType: DataType, hasVarArgs: Bool)
+
+  /// A pointer to a specified type.
   indirect case pointer(type: DataType)
+
+  /// An explicitly-sized array of a given type.
   indirect case array(field: DataType, length: Int?)
```
Comment (segiddins) [inline Sources/Base/Types.swift]: 🔥

File: Sources/Base/Types.swift (hunk 2 above)
Comment (harlanhaskins) [reply Sources/Base/Types.swift]: Good call

File: Sources/Import/ClangImporter.swift [hunk 4]
```diff
@@ -477,7 +477,9 @@ class ClangImporter: Pass {
       stringExpr.type = .pointer(type: .int8)
       return stringExpr
     case .number(let value, let raw):
-      return NumExpr(value: value, raw: raw, sourceRange: range)
+      let expr = NumExpr(value: value, raw: raw, sourceRange: range)
+      expr.type = .int64
```
Comment (segiddins) [inline Sources/Import/ClangImporter.swift]: will the handling for the suffixes overwrite this?

Comment (harlanhaskins) [reply Sources/Import/ClangImporter.swift]: Nope, it won't. Hmm... I'm going to rewrite this in terms of `Cursor.evaluate()` sometime soon

File: Sources/Semantic Analysis/ConstraintSolution.swift [hunk 5]
```diff
@@ -14,6 +14,9 @@ enum CoercionKind: Int {
   /// An explicit type variable has been reified to a concrete type.
   case genericPromotion
 
+  /// A stirng literal has been promoted to *Int8.
```
Comment (segiddins) [inline Sources/Semantic Analysis/ConstraintSolution.swift]: `string`

File: Sources/Semantic Analysis/ConstraintSolver.swift [hunk 6]
```diff
@@ -110,9 +119,46 @@ struct ConstraintSolver {
                               node: c.attachedNode,
                               caller: c.location)
 
-        return try solveSystem(system)
+        do {
+          return try solveSystem(system)
+        } catch {
+          break
```
Comment (segiddins) [inline Sources/Semantic Analysis/ConstraintSolver.swift]: 😭 this makes breakpoints on throws harder

File: Sources/Semantic Analysis/ConstraintSystem.swift [hunk 7]
```diff
@@ -77,7 +110,9 @@ struct ConstraintSystem {
       case let .equal(t1, t2):
         print("\(t1) == \(t2)")
       case let .conforms(t1, t2):
-        print("\(t1): \(t2)")
+        print("\(t1) :  \(t2)")
```
Comment (segiddins) [inline Sources/Semantic Analysis/ConstraintSystem.swift]: why the extra space?

File: Sources/Semantic Analysis/ConstraintSolution.swift (hunk 5 above)
Comment (harlanhaskins) [reply Sources/Semantic Analysis/ConstraintSolution.swift]: 😂

File: Sources/Semantic Analysis/OverloadResolver.swift [hunk 8]
```diff
@@ -114,12 +150,26 @@ struct OverloadResolver {
...
+  /// - Parameters:
+  ///   - expr: The subscript being resolved
+  ///   - candidates: The candidates through which to search.
+  /// - Returns: A resolution decision explaining exactly what was chosen by
+  ///            the overload system.
+  func resolve(_ expr: SubscriptExpr, candidates: [SubscriptDecl]) -> OverloadResolution<SubscriptDecl> {
+    return resolve(expr.args, candidates: candidates) { candidate in
+      expr.decl = candidate
+      csGen.visitSubscriptExpr(expr)
+      expr.decl = nil
```
Comment (segiddins) [inline Sources/Semantic Analysis/OverloadResolver.swift:199]: put this in a `defer`

File: Sources/Semantic Analysis/OverloadResolver.swift [hunk 9]
```diff
@@ -231,6 +278,11 @@ struct OverloadResolver {
       return .resolved(minSolutionCandidates[0].chosenDecl as! DeclType)
     }
 
+    for solution in minSolutionCandidates {
+      print("solution for \(solution.chosenDecl.formattedName)")
```
Comment (segiddins) [inline Sources/Semantic Analysis/OverloadResolver.swift]: debugging code

File: Sources/Semantic Analysis/Sema.swift [hunk 10]
```diff
@@ -134,30 +153,30 @@ class Sema: ASTTransformer, Pass {
   
   override func withScope(_ e: CompoundStmt, _ f: () -> Void) {
     let oldVarBindings = varBindings
+    let oldEnv = env
     super.withScope(e, f)
+    env = oldEnv
```
Comment (segiddins) [inline Sources/Semantic Analysis/Sema.swift]: `defer`

File: Sources/Semantic Analysis/Sema.swift [hunk 11]
```diff
@@ -168,26 +187,21 @@ class Sema: ASTTransformer, Pass {
...
     
-    if let rhs = decl.rhs, decl.typeRef == nil {
-      let type = rhs.type
-      guard rhs.type != .error else { return }
-      let canRhs = context.canonicalType(type)
+    if let rhs = decl.rhs {
+      let canRhs = context.canonicalType(rhs.type)
       if case .void = canRhs {
-        error(SemaError.incompleteTypeAccess(type: type, operation: "assign value from"),
+        error(SemaError.incompleteTypeAccess(type: canRhs,
```
Comment (segiddins) [inline Sources/Semantic Analysis/Sema.swift]: shouldn't this be the sugared type?

File: Sources/Semantic Analysis/ConstraintSystem.swift (hunk 7 above)
Comment (harlanhaskins) [reply Sources/Semantic Analysis/ConstraintSystem.swift]: ¯\_(ツ)_/¯

**Comment by segiddins** [comment]
To summarize the state of this PR:
Harlan and I have discussed how to fix the regressions, and the solution we've converged upon is that type solving needs to be done at the _statement_, rather than _expression_, level, to allow for cases in which expressions have sub-expressions with multiple possible types -- (see for example what happens with ternaries). For this to happen, the overload resolution code has to be juggled about to play nicely with expressions not being fully types until the enclosing statement has been fully visited.

**Comment by CodaFi** [comment]
> type solving needs to be done at the statement, rather than expression, level, to allow for cases in which expressions have sub-expressions with multiple possible types -- (see for example what happens with ternaries)

The rule for this is that you type check both branches and if they yield two disjunctions you can either force the user to specify a contextual type or have some merge heuristic (try to take an intersection and see if you get a singular set, maybe?).  Otherwise you can use whatever branch that doesn't have a disjunct in it as a contextual type and keep solving.

Solving at the statement level should only be done if you intend to start propagating solution sets *between statements*.

**Pull request closed without merging**